exa/core/data.py

```diff
@@ -4,11 +4,182 @@
+            except Exception as e:
+                print("attempt to import source failed")
+                source = None
+        return cls(source=source, **cfg)
+
+    def data(self, df=None, cache=True):
+        """Return the currently stored data in the
+        Data object. If df is provided, store that
+        as the current data and return it. Otherwise,
+        determine the provider to execute to obtain
+        the data, store it and return it.
+
+        Note:
+            force re-evaluation of source if cache is False
+        """
+        _data = getattr(self, '_data', None)
+        if not cache:
+            _data = None
+        if df is not None:
+            _data = df
+        if _data is None and self.source is not None:
+            _data = self.source(
+                *self.source_args, **self.source_kws
+            )
+        self._data = self._validate_data(_data)
+        return self._data
+
+    def _validate_data(self, df):
+        if not isinstance(df, pd.DataFrame):
+            self.log.warning("data not a dataframe, skipping validation")
+            return df
+        # set index name, support set index?
+        # or a (multi-col) uniqueness constraint?
+        if self.index and df.index.name != self.index:
+            self.log.debug("setting index name {}".format(self.index))
+            df.index.name = self.index
+        # guarantee existence
+        missing = set(self.columns).difference(df.columns)
+        if missing:
+            raise RequiredColumnError(missing, self.name)
+        return self._set_categories(df)
+
+    def _set_categories(self, df, reverse=False):
+        """For specified categorical fields,
+        convert to pd.Categoricals.
+
+        Note:
+            If reverse is True, revert categories
+        """
+        for col, typ in self.categories.items():
+            if col in df.columns:
+                conv = {True: typ, False: 'category'}[reverse]
+                df[col] = df[col].astype(conv)
+            else:
+                self.log.debug(
+                    "categorical {} specified but not in data".format(col)
+                )
+        return df
+
+
+def load_isotopes():
+    """Minimal working example of a pluggable
+    callable to serve as a data provider in the
+    Data API.
+    """
+    path = exa.cfg.resource('isotopes.json')
+    df = pd.read_json(path, orient='values')
+    df.columns = ('A', 'Z', 'af', 'afu',
+                  'cov_radius', 'van_radius', 'g',
+                  'mass', 'massu', 'name', 'eneg',
+                  'quad', 'spin', 'symbol', 'color')
+    return df.sort_values(by=['symbol', 'A']).reset_index(drop=True)
+
+class Isotopes(Data):
+    """An isotopes database data object.
+
+    .. code-block:: python
+
+        import exa
+        df = exa.Isotopes().data()
+    """
+
+    def __init__(self, *args, **kws):
+        source = kws.pop('source', load_isotopes)
+        super().__init__(*args, source=source, **kws)
```

Comment:
This appears twice - boilerplate for derived `Data` classes? Can it be abstracted somehow?

Reply:
Perhaps I was too eager in subclassing `Data`. I believe the following is functionally equivalent:
`exa.Data(source=load_isotopes, name='isotopes')` and similar goes for `Constants`. Bad habit, perhaps.

Reply:
That usage is definitely intuitive. For subclassing how do make sure there is no additional required boilerplate? And I'm totally for subclassing - showing am example of or otherwise.

Reply:
Well we get `name` based on subclass name 😄 but really it might depend on the attribute.  I think `source` is definitely the linchpin in the `Data` class so it may require additional attention.

However, I think in principle we would like to limit the *need* for subclassing `Data`.